File: platform/ios/SeckShare/Sources/WGClient.swift
// Plan-17 iOS WG client scaffold.
//
// Holds the desktop's PairingBundle (loaded from the app's shared
// container after the user completed pairing). For each analyze call,
// brings up a WireGuard tunnel via BoringTun, sends a JSON-RPC
// `seck.analyze` over the tunnel, returns the report.
//
// The actual BoringTun glue is left to the executor — multiple Swift
// wrappers of BoringTun exist (cloudflare/boringtun has a `boringtun`
// crate that compiles to a static lib + Swift FFI). Pinning one is out
// of scope for this scaffold.
import Foundation

struct PairingBundle: Codable {
    let host_public_hex: String
    let psk_hex: String
    let host_endpoint: String
    let fingerprint_sha3_256: String
}

final class WGClient {
    static let shared = WGClient()

    private init() {}

    /// Load the bundle persisted by the host pairing flow. The user
    /// scans the desktop QR; the iOS UI parses + verifies fingerprint;
    /// the bundle JSON is written to the App Group container.
    func loadBundle() -> PairingBundle? {
        guard let dir = FileManager.default.containerURL(
            forSecurityApplicationGroupIdentifier: "group.net.seck.share") else { return nil }
        let url = dir.appendingPathComponent("pairing.json")
        guard let data = try? Data(contentsOf: url) else { return nil }
        return try? JSONDecoder().decode(PairingBundle.self, from: data)
    }

    /// Submit `contents` to the paired host. Calls back with the JSON
    /// report (or an error string). Network errors and pairing-missing
    /// produce user-facing strings rather than throwing — the share
    /// extension UI displays whatever this returns.
    func analyze(filename: String,
                 contents: Data,
                 completion: @escaping (String) -> Void) {
        guard let bundle = loadBundle() else {
            completion("Not paired — run `seck pair` on your desktop, then scan the QR.")
            return
        }
        // Build JSON-RPC frame
        let frame: [String: Any] = [
            "jsonrpc": "2.0",
            "id": UUID().uuidString,
            "method": "seck.analyze",
            "params": [
                "filename": filename,
                "content_base64": contents.base64EncodedString(),
            ]
        ]
        guard let body = try? JSONSerialization.data(withJSONObject: frame) else {
            completion("Failed to encode request.")
            return
        }
        // BoringTun-glued UDP transport goes here in the executor build.
        // For the scaffold, we acknowledge the pairing and the encoded
        // request size so the UI flow is end-to-end demonstrable.
        let preview = "Encoded \(body.count) bytes; would send to \(bundle.host_endpoint)."
        completion(preview)
    }
}

File: platform/android/seckshare/src/main/java/net/seck/share/WGClient.kt
// Plan-17 Android WG client scaffold.
//
// Mirrors the iOS WGClient.swift: load PairingBundle (from a file in
// the app's private storage, written by the pairing screen), construct
// JSON-RPC `seck.analyze`, send over WireGuard (BoringTun JNI), return
// the report.
package net.seck.share

import android.content.Context
import android.util.Base64
import org.json.JSONObject
import java.io.File
import java.util.UUID
import kotlin.concurrent.thread

object WGClient {

    data class PairingBundle(
        val hostPublicHex: String,
        val pskHex: String,
        val hostEndpoint: String,
        val fingerprintSha3_256: String,
    )

    private fun loadBundle(ctx: Context): PairingBundle? {
        val f = File(ctx.filesDir, "pairing.json")
        if (!f.exists()) return null
        val o = JSONObject(f.readText())
        return PairingBundle(
            hostPublicHex = o.getString("host_public_hex"),
            pskHex = o.getString("psk_hex"),
            hostEndpoint = o.getString("host_endpoint"),
            fingerprintSha3_256 = o.getString("fingerprint_sha3_256"),
        )
    }

    fun analyze(
        ctx: Context,
        filename: String,
        contents: ByteArray,
        callback: (String) -> Unit,
    ) {
        val bundle = loadBundle(ctx)
        if (bundle == null) {
            callback("Not paired — run `seck pair` on your desktop, then scan the QR.")
            return
        }
        thread {
            val frame = JSONObject().apply {
                put("jsonrpc", "2.0")
                put("id", UUID.randomUUID().toString())
                put("method", "seck.analyze")
                put("params", JSONObject().apply {
                    put("filename", filename)
                    put("content_base64", Base64.encodeToString(contents, Base64.NO_WRAP))
                })
            }
            val body = frame.toString().toByteArray(Charsets.UTF_8)
            // BoringTun JNI tunnel goes here in the executor build.
            callback("Encoded ${body.size} bytes; would send to ${bundle.hostEndpoint}.")
        }
    }
}
